cpp/src/utilities/device_operators.cuh

```diff
@@ -24,103 +24,130 @@
  * ---------------------------------------------------------------------------**/
 
 #include <cudf/cudf.h>
-#include <utilities/cudf_utils.h>       // need for CUDA_HOST_DEVICE_CALLABLE
-#include <cudf/utilities/legacy/wrapper_types.hpp>
+#include <utilities/cudf_utils.h>  // need for CUDA_HOST_DEVICE_CALLABLE
 #include <cudf/utilities/error.hpp>
+#include <cudf/utilities/legacy/wrapper_types.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <type_traits>
 
 namespace cudf {
 
 // ------------------------------------------------------------------------
 // Binary operators
 /* @brief binary `sum` operator */
 struct DeviceSum {
-    template<typename T>
-    CUDA_HOST_DEVICE_CALLABLE
-    T operator() (const T &lhs, const T &rhs) {
-        return lhs + rhs;
-    }
-
-    template<typename T>
-    static constexpr T identity() { return T{0}; }
+  template <typename T,
+            typename std::enable_if_t<cudf::is_timestamp<T>()>* = nullptr>
+  CUDA_HOST_DEVICE_CALLABLE T operator()(const T& lhs, const T& rhs) {
+    return T{DeviceSum{}(lhs.time_since_epoch(), rhs.time_since_epoch())};
+  }
+
+  template <typename T,
+            typename std::enable_if_t<!cudf::is_timestamp<T>()>* = nullptr>
+  CUDA_HOST_DEVICE_CALLABLE T operator()(const T& lhs, const T& rhs) {
+    return lhs + rhs;
+  }
+
+  template <typename T>
+  static constexpr T identity() {
+    return T{0};
+  }
 };
 
 /* @brief `count` operator - used in rolling windows */
 struct DeviceCount {
-    template<typename T>
-    CUDA_HOST_DEVICE_CALLABLE
-    T operator() (const T &, const T &rhs) {
-        return rhs + T{1};
-    }
-
-    template<typename T>
-    static constexpr T identity() { return T{0}; }
+  template <typename T,
+            typename std::enable_if_t<cudf::is_timestamp<T>()>* = nullptr>
+  CUDA_HOST_DEVICE_CALLABLE T operator()(const T& lhs, const T& rhs) {
+    return T{DeviceCount{}(lhs.time_since_epoch(), rhs.time_since_epoch())};
```

Comment:
This seems to implement sum, not count.

Reply:
This is delegating to the other overload of the `DeviceCount` functor, but has to unwrap both arguments since its argument list is `(const T& lhs, const T& rhs)`

Reply:
I misread this.